File: lessons/03_associativity/util/unique_set.lua
-- Unique set
-- You can ask this class for the id of any element. Elements will be numbered from 1..n, the number of distinct elements that has been seens by this class.
-- This can be used to make sure that only unique references to constants or strings or tables are stored

unique_set_mt = {}

unique_set = {}
unique_set_mt = { __index = unique_set }

function create_unique_set()
	local set = {}
	setmetatable(set, unique_set_mt)

	set.values = {}
	set.count = 0

	return set
end

function unique_set:get_id(key)
	local id = self.values[key]

	if (id==nil) then
		self.count = self.count + 1
		id = self.count
		self.values[key] = id
	end

	return id
end

function unique_set:to_table()
	local t = {}

	for i,v in pairs(self.values) do
		t[v] = i
	end

	return t
end

File: java_port/06_conditionals+loops/src/main/java/compilertalk/util/UniqueSet.java
package compilertalk.util;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

/**
 * Unique set.
 * <ul>
 * <li>You can ask this class for the id of any element. Elements will be
 * numbered from 1..n, the number of distinct elements that has been seens by
 * this class.</li>
 * <li>This can be used to make sure that only unique references to constants or
 * strings or tables are stored.</li>
 * </ul>
 */
public class UniqueSet<T> {

	private final Map<T, Short> values = new HashMap<T, Short>();
	private short count = 0;

	public short getIDZeroBased(T key) {
		Short id = values.get(key);
		if (id == null) {
			id = count;
			count++;
			values.put(key, id);
		}
		return id;
	}

	/**
	 * @see List#toArray(Object[])
	 */
	public T[] toArray(T[] a) {
		if (a.length < values.size())
			a = Arrays.copyOf(a, values.size());
		for (Entry<T, Short> i_v : values.entrySet()) {
			T i = i_v.getKey();
			Short v = i_v.getValue();
			a[v] = i;
		}
		if (a.length > values.size())
			a[values.size()] = null;
		return a;
	}
}
